head.lisp

```diff
@@ -40,22 +40,29 @@
                                 (parse-error ()
                                   nil))))
 
+(defun list-heads ()
+  "Return a list of HEAD structures without duplicates"
+  (delete-duplicates
+   (loop for line in (split-string (run-shell-command "xdpyinfo -ext XINERAMA" t))
+         for head = (parse-xinerama-head line)
+         when head
+           collect head)
+   :test #'frames-overlap-p))
+
+(defun number-heads (heads)
+  "Number each head according to its position in the list."
+  (let ((head-index -1))
+    (flet ((set-head-number (head)
+             (setf (head-number head)
+                   (incf head-index))))
+      (mapc #'set-head-number heads))))
+
```

Comment:
Honestly the flet+incf+mapc seems like scratching your nose behind your head to keep it functional.  The operation is inherently declarative and its easier to mentally parse the loop version.